olp-cpp-sdk-dataservice-read/src/StreamLayerClientImpl.cpp

```diff
@@ -21,23 +21,142 @@
 
 #include <olp/core/cache/DefaultCache.h>
 #include <olp/core/client/OlpClientSettingsFactory.h>
+#include <olp/core/logging/Log.h>
+#include <olp/core/thread/TaskScheduler.h>
+
+#include "ApiClientLookup.h"
+#include "Common.h"
+#include "generated/api/StreamApi.h"
 
 namespace olp {
 namespace dataservice {
 namespace read {
 
+namespace {
+constexpr auto kLogTag = "StreamLayerClientImpl";
+constexpr auto kSubscriptionModeParallel = "parallel";
+constexpr auto kSubscriptionModeSerial = "serial";
+constexpr auto kStreamService = "stream";
+constexpr auto kStreamVersion = "v2";
+
+boost::optional<std::string> GetSubscriptionMode(
+    SubscribeRequest::SubscriptionMode mode) {
+  if (mode == SubscribeRequest::SubscriptionMode::kParallel) {
+    return kSubscriptionModeParallel;
+  } else if (mode == SubscribeRequest::SubscriptionMode::kSerial) {
+    return kSubscriptionModeSerial;
+  }
+  return boost::none;
+}
+
+}  // namespace
+
 StreamLayerClientImpl::StreamLayerClientImpl(client::HRN catalog,
                                              std::string layer_id,
                                              client::OlpClientSettings settings)
     : catalog_(std::move(catalog)),
       layer_id_(std::move(layer_id)),
-      settings_(std::move(settings)) {
+      settings_(std::move(settings)),
+      pending_requests_(std::make_shared<client::PendingRequests>()) {
   if (!settings_.cache) {
     settings_.cache = client::OlpClientSettingsFactory::CreateDefaultCache({});
   }
 }
 
-StreamLayerClientImpl::~StreamLayerClientImpl() = default;
+StreamLayerClientImpl::~StreamLayerClientImpl() {
+  pending_requests_->CancelAllAndWait();
+}
+
+bool StreamLayerClientImpl::CancelPendingRequests() {
+  OLP_SDK_LOG_TRACE(kLogTag, "CancelPendingRequests");
+  return pending_requests_->CancelAll();
+}
+
+client::CancellationToken StreamLayerClientImpl::Subscribe(
+    SubscribeRequest request,
+    SubscribeRequest::SubscribeResponseCallback callback) {
+  using SubscribeResponse =
+      client::ApiResponse<model::SubscribeResponse, client::ApiError>;
+
+  auto self = shared_from_this();
+
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (subscription_id_) {
+      lock.unlock();
+      OLP_SDK_LOG_WARNING_F(kLogTag, "Subscribe : already subscribed");
```

Comment:
```suggestion
      OLP_SDK_LOG_WARNING_F(kLogTag, "Subscribe, id=%s: already subscribed", request.GetSubscriptionId().get().c_str());
```
Please, provide subscription_id in logs in order for log message to be more informative